pkg/pool-weighted/contracts/managed/ManagedPool.sol

```diff
@@ -587,6 +710,73 @@ contract ManagedPool is ManagedPoolSettings {
         return ComposablePoolLib.dropBpt(registeredTokens, registeredBalances);
     }
 
+    // Circuit Breakers
+
+    /**
+     * @dev Get the current BPT price, given its components (e.g., the end state of an operation).
+     */
+    function _getBptPrice(
+        uint256 virtualSupply,
+        uint256 normalizedWeight,
+        uint256 upscaledBalance
+    ) private pure returns (uint256) {
+        return virtualSupply.mulUp(normalizedWeight).divDown(upscaledBalance);
+    }
+
+    /**
+     * @dev Trip the circuit breaker if the current BPT price is less than the lower bound, or greater than the
+     * upper bound. For performance reasons, check for a zero bound (= no circuit breaker set) externally,
+     * avoiding computation of the current BPT price when no breaker is set.
+     */
+    function _checkCircuitBreaker(
+        uint256 bptPriceBound,
+        uint256 virtualSupply,
+        uint256 normalizedWeight,
+        uint256 upscaledBalance,
+        bool isLowerBound
+    ) private pure {
+        uint256 currentBptPrice = _getBptPrice(virtualSupply, normalizedWeight, upscaledBalance);
+
+        _require(
+            isLowerBound ? currentBptPrice >= bptPriceBound : currentBptPrice <= bptPriceBound,
+            Errors.CIRCUIT_BREAKER_TRIPPED
+        );
+    }
```

Comment:
We're essentially reimplementing `hasCircuitBreakerTripped` here. I think a better route would be to have a single sided variant of that function as ManagedPool shouldn't really need to know what a "bptPrice" is.

Reply:
Yes - I noticed I wasn't using that function, and was thinking along the same lines (better encapsulation).